### PR TITLE
[SMALLFIX] Added heartbeat check in BlockMaster

### PR DESCRIPTION
**Pull request by yupeng9**

### CODE REVIEW

**Comment by gpang** [comment]
Thanks for this patch! I have a high level question. What do you think about having each master (in MasterBase) owning the executor service for its own threads, instead of the TachyonMaster?

**Comment by yupeng9** [comment]
That's a good question. There're tradeoffs between having a single service (ease of thread pool management etc) and one per master (flexibility, e.g. thread pool naming). I don't have preference on either. What do you think?

**Comment by gpang** [comment]
Hrmmm, let's go with having each master own its own executor service (in master base), since each master would have a better idea on how many threads it will need, and can size the executor service appropriately. And maybe it will be easier to name the threads more appropriately.

**Comment by yupeng9** [comment]
Cool. SGTM.

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java [hunk 1]
```diff
@@ -69,6 +78,12 @@
    */
   private final BlockIdGenerator mBlockIdGenerator = new BlockIdGenerator();
   private final Set<Long> mLostBlocks = new HashSet<Long>();
+  private final BlockingQueue<MasterWorkerInfo> mLostWorkers =
+      new LinkedBlockingQueue<MasterWorkerInfo>();
+  private final TachyonConf mTachyonConf;
+
+  // heart beat service
+  private Future<?> mHeartbeat;
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: This is the heartbeat for detecting lost workers, so it the comments should reflect that, and maybe a more descriptive variable name.

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java [hunk 2]
```diff
@@ -146,7 +163,9 @@ public void writeToJournal(JournalOutputStream outputStream) throws IOException
   public void start(boolean asMaster) throws IOException {
     startMaster(asMaster);
     if (isMasterMode()) {
-      // TODO: start periodic heartbeat threads.
+      mHeartbeat = getExecutorService()
+          .submit(new HeartbeatThread("Master Heartbeat", new MasterInfoHeartbeatExecutor(),
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: I think the name should be more descriptive like "Lost Worker Detector" or something.

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java [hunk 3]
```diff
@@ -449,4 +470,45 @@ private void processWorkerAddedBlocks(MasterWorkerInfo workerInfo,
       }
     }
   }
+
+  /**
+   * Master info periodical status check.
+   */
+  public final class MasterInfoHeartbeatExecutor implements HeartbeatExecutor {
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: I don't think "MasterInfoHeartbeatExecutor" is very descriptive. Maybe "LostWorkerDetectorHeartbeat" or something like that?

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java [hunk 4]
```diff
@@ -449,4 +470,45 @@ private void processWorkerAddedBlocks(MasterWorkerInfo workerInfo,
...
+    @Override
+    public void heartbeat() {
+      LOG.debug("System status checking.");
+
+      Set<Long> lostWorkers = new HashSet<Long>();
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: Why is this set necessary?

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java [hunk 5]
```diff
@@ -449,4 +470,45 @@ private void processWorkerAddedBlocks(MasterWorkerInfo workerInfo,
...
+
+      synchronized (mWorkers) {
+        for (MasterWorkerInfo worker : mWorkers) {
+          int masterWorkerTimeoutMs =
+              mTachyonConf.getInt(Constants.MASTER_WORKER_TIMEOUT_MS);
+          if (CommonUtils.getCurrentMs()
+              - worker.getLastUpdatedTimeMs() > masterWorkerTimeoutMs) {
+            LOG.error("The worker " + worker + " got timed out!");
+            mLostWorkers.add(worker);
```
Comment (gpang) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: When do workers leave mLostWorkers?

**Comment by gpang** [comment]
I left some comments. Thanks for the PR!

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java (hunk 1 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: Sounds good.

File: servers/src/main/java/tachyon/master/next/block/BlockMaster.java (hunk 5 above)
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/next/block/BlockMaster.java]: Good catch!

**Comment by gpang** [comment]
LGTM

**Pull request merged**